### ISSUE TITLE
[custom-builds-rc] globbing for module types / redo `upload_format` configuration

### ISSUE THREAD

**Issue by xortive** [cloudflare/wrangler#1841]
Right now, the modules `upload-format` in custom builds mode has the following behavior:

- files in `upload_dir` are considered by the following rules
- `upload_dir` must exist, and must not be the project root.
- Files in `upload_exclude` are ignored by the following rules
- Only files in `upload_include` are considered by the following rules
- If `upload_include` and `upload_exclude` are both specified, `upload_include` is used as it is more specific.
- Files with known extensions are uploaded as a specific module type
- Files with that are left over after the above rules, are an error.

We should replace this behavior with file globs for each module type, and remove `upload_include` and `upload_exclude`. `upload_dir` will be kept, to avoid having to specify the path to the upload dir for each module. The default value for `upload_dir` will still be `./dist`.

Additionally, We should replace `upload_format` with an [internally tagged enum union](https://serde.rs/enum-representations.html#internally-tagged) called `upload`.

modules configuration will now look as follows

```toml
[build]
command = "npm r build"
[build.upload]
format = "modules"
dir = "dist"
# determines the prefix path for all of the module resolution rules below.
main = "./index.mjs"
# Must be a relative path inside of dist, including `./`
# The module type is determined by the rules below. Currently, we only allow ES modules to be the main module

# The default values for each module type are listed below.
esm = ["*.mjs"]
cjs = ["*.js", "*.cjs"]
text = ["*.txt"]
data = ["*.bin"]
compiled_wasm = ["*.wasm"] # NOT a default, see note below.
```

> Our wasm module implementation is non-standard, we should require you to specify that you are using `wasm_compiled`, so if we introduce standard wasm modules in the future, we are not stuck with a non-standard default. Most folks using wasm, will be using a template to get started, and this will be set inside of the template.

service-worker mode will look as follows

```toml
[build]
command = "npm r build"
upload.format = "service-worker"
```

**Comment by xortive** [comment]
I realized the above solution doesn't have a nice way to resolve conflicts, where two module types's globs match the same file. I found a nice solution to this: make module type rules an array of tables. each rule set is evaluated in order, and conflicts within a ruleset are prohibited.

each rule set contains a key for each module type, and the value can either be an array of globs (shorthand for a rule with `fallthrough = false`), or a rule object. This will be implemented using serde's untagged unions feature.

`fallthrough = false` is the default, since `fallthrough = true` as a default could cause breakage if we change the default rule set, so we should make users opt-in to potential changes.

```toml
[build]
command = "npm r build"
[build.upload]
format = "modules"
dir = "dist"
# determines the prefix path for all of the module resolution rules below.
main = "./index.mjs"
# Must be a relative path inside of dist, including `./`
# The module type is determined by the rules below.
# Currently, we only allow ES modules to be the main module

[[build.upload.rules]]
esm.globs = ["bar.js", "foo.js"]
esm.fallthrough = true # if this rule isn't matched, fall through
data = ["*.jpg"] # overrides the default `data` rule.

[[build.upload.rules]]
# This entry is implied to be the last entry by default.
esm.globs = ["*.mjs"]
cjs.globs = ["*.js", "*.cjs"]
text.globs = ["*.txt"]
data.globs = ["*.bin"]
# wasm is intentionally omitted, since we don't want to have our non-standard
# wasm module implementation be a default forever.
```

**Comment by a-robinson** [comment]
Mostly makes sense to me.

I don't quite get the `fallthrough` bit, though. If the last table is always implied by default, isn't everyone becoming dependent on the default ruleset whenever they don't use overrides?

And it's a little weird to have to specify `esm.globs` if you want to also set `esm.fallthrough` compared to just using `esm = [...]` if you don't care to set `fallthrough`.

You may want to have others look at this too, though. The array-of-tables approach strikes me as a little complex, but I suppose that more than one table will rarely be needed and the option is mostly there to ensure power users can do what they want.

An inverted alternative would be to provide an ordered set of globs, something like:

```
rules = [
  ["bar.js", "foo.js"] => esm,
  ["*.jpg"] => data,
  ["*.mjs"] => esm,
  ...
]
```

But that'd probably end up looking pretty ugly when made to be work as toml.

**Comment by xortive** [comment]
> I don't quite get the fallthrough bit, though. If the last table is always implied by default, isn't everyone becoming dependent on the default ruleset whenever they don't use overrides?

(I assume by `override` you mean a `[[build.upload.rules]]` section.)
Yeah, I guess it doesn't protect 100% against dependency on the default rules, but it makes it clearly obvious if you are using them.
The decision to make `fallthrough = false` was guided by switch statements in C/C++, where fallthrough is the default and can come with weird bugs. I wanted to make it obvious when further rules are being considered.

> And it's a little weird to have to specify esm.globs if you want to also set esm.fallthrough compared to just using esm = [...] if you don't care to set fallthrough.

you could also do `esm = { fallthrough = true, globs = [...]}`, the problem is toml inline tables have to be a single line. Both will work, it's up to us which one we'll use in examples.

> You may want to have others look at this too, though. The array-of-tables approach strikes me as a little complex, but I suppose that more than one table will rarely be needed and the option is mostly there to ensure power users can do what they want.

The defaults should work just fine for the vast majority of issues. I think this will come into play with existing complex builds, when trying to use tooling meant for other JS environments that don't use `.mjs`, and until wasm ES modules are standardized, when using wasm. I suspect more than one table will be extremely rare, but the implementation is the same for one override table or many.

> An inverted alternative would be to provide an ordered set of globs, but that'd probably end up looking pretty ugly when made to be work as toml.

it's not so bad as inline array/table syntax, but multi-line inline tables are not allowed (well, you could have a line break inside the array, but it's strongly discouraged by the toml spec...)

```toml
rules = [
   { type = "esm", fallthrough = true, globs = ["bar.js, foo.js"] },
   { type = "data", globs = ["*.jpg"] }
   # ... implied rules go here
]
```

As array of tables syntax, it's kinda verbose.

```toml
[[build.upload.rules]]
type = "esm"
fallthrough = true
globs = ["bar.js, foo.js"]

[[build.upload.rules]]
type = "data"
globs = ["*.jpg"]
```

You'd completely eliminate the chance of having conflicts where multiple module types' globs match a single file, since an entry in the list only handles one module type. You can easily just make a new ruleset in my proposed syntax when this happens though.

**Comment by a-robinson** [comment]
I'm on board with your thinking 👍 

The question I forgot to ask, though, is what happens to existing config files that people have? This doesn't break anyone other than early custom builds users, does it?

**Comment by xortive** [comment]
> I'm on board with your thinking

did you have a preference between the `array of type -> glob mappings` vs `array of rulesets (my original proposal)`

> The question I forgot to ask, though, is what happens to existing config files that people have? This doesn't break anyone other than early custom builds users, does it?

breaks existing custom builds users. I'll add a check for `upload_format` in the final version of the PR that explains how to migrate.

**Comment by a-robinson** [comment]
> did you have a preference between the array of type -> glob mappings vs array of rulesets (my original proposal)

Your original proposal

**Comment by Electroid** [comment]
For future reference, after some trade-off deliberation and discussion we ended up going with the alternative proposal. Generally, it's cleaner and more compact. 
```toml
rules = [
   { type = "esm", fallthrough = true, globs = ["bar.js, foo.js"] },
   { type = "data", globs = ["*.jpg"] }
   # ... implied rules go here
]
```

**Comment by ObsidianMinor** [comment]
The alternative proposal almost exactly like the first one but with loosely typed "type" fields. Consider an alternative way to write it:
```toml
[[rules]]
type = "esm"
fallthrough = true
globs = ["bar.js", "foo.js"]
```
At this point why not make it another table
```toml
[[rules]]
esm.fallthrough = true
esm.globs = ["bar.js", "foo.js"]
```
Feels like we've effectively just introduced a type field for type field's sake.

**Comment by xortive** [comment]
@ObsidianMinor it's unclear to me how you would write serde structs for the second variant. the first is trivial to implement using https://serde.rs/enum-representations.html#internally-tagged

**Comment by ObsidianMinor** [comment]
```rust
struct TypeRule {
    pub fallthrough: bool,
    pub globs: Vec<String>,
}

struct Rule {
    pub esm: Option<TypeRule>,
    pub data: Option<TypeRule>,
}

struct Modules {
    // some other fields...
    pub rules: Vec<Rule>,
}
```
Unless we're going with one rule type per table (not sure why), we use the default enum representation "externally tagged"
```
#[serde(rename_all = "lowercase")]
enum Rule {
    Esm(TypeRule),
    Data(TypeRule),
}
```

**Comment by xortive** [comment]
> one rule type per table 

that was the intention, yes

**Comment by Electroid** [comment]
Released in 1.16 🎉
